compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Context.hs

```diff
@@ -169,69 +243,247 @@ data UpdChoice = UpdChoice
-    { _uscgUpdate :: ![Cond Upd]
-      -- ^ The list of updates.
-    , _uscgChoice :: ![Cond UpdChoice]
-      -- ^ The list of exercised choices.
-    } -> UpdateSet 'ChoiceGathering
-  UpdateSetS ::
-    { _ussUpdate :: ![Cond Upd]
-      -- ^ The list of updates.
-    } -> UpdateSet 'Solving
-
-class GenPhase ph where
-  emptyUpdateSet :: UpdateSet ph
+-- | Class containing the environment, and operations on it, for each generator phase.
+class IsPhase (ph :: Phase) where
+  -- | The updates which can be performed.
+  data Upd ph
+  -- | The environment for the DAML-LF verifier.
+  data Env ph
+  -- | Construct a base update.
+  baseUpd :: Rec [Cond BaseUpd] -> Upd ph
+  -- | Construct a choice exercise update.
+  choiceUpd :: Cond UpdChoice -> Upd ph
+  -- | Construct a value update.
+  valueUpd :: Cond (Qualified ExprValName) -> Upd ph
+  -- | Map over a single base update.
+  mapBaseUpd :: (Rec [Cond BaseUpd] -> Upd ph) -> Upd ph -> Upd ph
+  -- | Check whether the update set contains any choice references.
+  containsChoiceRefs :: UpdateSet ph -> Bool
+  -- | Extend the conditional of an update.
+  extCondUpd :: BoolExpr -> Upd ph -> UpdateSet ph
+  -- | Construct an empty environment.
   emptyEnv :: Env ph
-
-instance GenPhase 'ValueGathering where
-  emptyUpdateSet = UpdateSetVG [] [] []
+  -- | Combine two environments.
+  concatEnv :: Env ph -> Env ph -> Env ph
+  -- | Get the skolemised term variables and fields from the environment.
+  envSkols :: Env ph -> [Skolem]
+  -- | Update the skolemised term variables and fields in the environment.
+  setEnvSkols :: [Skolem] -> Env ph -> Env ph
+  -- | Get the bound values from the environment.
+  envVals :: Env ph -> HM.HashMap (Qualified ExprValName) (Expr, UpdateSet ph)
+  -- | Update the bound values in the environment.
+  setEnvVals :: HM.HashMap (Qualified ExprValName) (Expr, UpdateSet ph) -> Env ph -> Env ph
+  -- | Get the data constructors from the environment.
+  envDats :: Env ph -> HM.HashMap TypeConName DefDataType
+  -- | Update the data constructors in the environment.
+  setEnvDats :: HM.HashMap TypeConName DefDataType -> Env ph -> Env ph
+  -- | Get the fetched cid's mapped to their current variable name, along with
+  -- a list of any potential old variable names, from the environment.
+  envCids :: Env ph -> HM.HashMap Cid (ExprVarName, [ExprVarName])
+  -- | Update the fetched cid's in the environment.
+  setEnvCids :: HM.HashMap Cid (ExprVarName, [ExprVarName]) -> Env ph -> Env ph
+  -- | Get the set of preconditions from the environment.
+  envPreconds :: Env ph -> HM.HashMap (Qualified TypeConName) (Expr -> Expr)
+  -- | Update the set of preconditions in the environment.
+  setEnvPreconds :: HM.HashMap (Qualified TypeConName) (Expr -> Expr) -> Env ph -> Env ph
+  -- | Get the additional constraints from the environment.
+  envCtrs :: Env ph -> [BoolExpr]
+  -- | Update the additional constraints in the environment.
+  setEnvCtrs :: [BoolExpr] -> Env ph -> Env ph
```

Comment:
I don't think that's correct. It could also be called through `genChoice` -> `extEnvContract` -> `extCtr` in the `ValueGathering` phase.
However, as this function can't throw any errors, I think that's less of an issue.